File: creational/builder/dynamic_workflow/typescript/dynamic_workflow.test.ts
import { DynamicWorkflowBuilder, DynamicWorkflow } from './dynamic_workflow';

// Mock console methods to capture output
const consoleLogSpy = jest.spyOn(console, 'log').mockImplementation();
const consoleWarnSpy = jest.spyOn(console, 'warn').mockImplementation();
const consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation();

describe('DynamicWorkflowBuilder', () => {

    beforeEach(() => {
        // Clear mocks before each test
        consoleLogSpy.mockClear();
        consoleWarnSpy.mockClear();
        consoleErrorSpy.mockClear();
    });

    test('should build an empty workflow', () => {
        const builder = new DynamicWorkflowBuilder();
        const workflow = builder.build();
        expect(workflow).toBeInstanceOf(DynamicWorkflow);
        expect(workflow.getSteps()).toHaveLength(0);
        expect(consoleWarnSpy).toHaveBeenCalledWith("Builder Warning: Building an empty workflow.");
    });

    test('should add a single step', () => {
        const builder = new DynamicWorkflowBuilder();
        const workflow = builder.addStep("send_email", { to: "test@example.com" }).build();
        const steps = workflow.getSteps();
        expect(steps).toHaveLength(1);
        expect(steps[0].type).toBe("send_email");
        expect(steps[0].params).toEqual({ to: "test@example.com" });
    });

    test('should add multiple steps with chaining', () => {
        const builder = new DynamicWorkflowBuilder();
        const workflow = builder
            .addStep("run_script", { path: "/script1" })
            .addStep("notify_slack", { channel: "#test" })
            .build();
        const steps = workflow.getSteps();
        expect(steps).toHaveLength(2);
        expect(steps[0].type).toBe("run_script");
        expect(steps[1].type).toBe("notify_slack");
        expect(steps[1].params).toEqual({ channel: "#test" });
    });

    test('addStep should return the builder instance (this)', () => {
        const builder = new DynamicWorkflowBuilder();
        const returnValue = builder.addStep("test_step");
        expect(returnValue).toBe(builder);
    });
});

describe('DynamicWorkflow Execution', () => {

     beforeEach(() => {
        // Clear mocks before each test
        consoleLogSpy.mockClear();
        consoleWarnSpy.mockClear();
        consoleErrorSpy.mockClear();
    });

    test('should execute an empty workflow', () => {
        const workflow = new DynamicWorkflow([]);
        workflow.execute();
        // Check if the specific message for empty workflow was logged
        expect(consoleLogSpy).toHaveBeenCalledWith("Workflow has no steps.");
    });

    test('should execute known steps successfully', () => {
        const steps = [
            { type: "send_email", params: { to: "exec@test.com", subject: "Exec Test" } },
            { type: "run_script", params: { path: "/exec/test.py" } },
            { type: "notify_slack", params: { channel: "#exec", message: "Done" } }
        ];
        const workflow = new DynamicWorkflow(steps);
        workflow.execute();

        // Check if log messages contain expected simulation output
        const logOutput = consoleLogSpy.mock.calls.flat().join('\n');
        expect(logOutput).toContain("Simulating sending email to 'exec@test.com'");
        expect(logOutput).toContain("Simulating running script: '/exec/test.py'");
        expect(logOutput).toContain("Simulating notifying Slack channel '#exec'");

        // Check that no warnings or errors were logged
        expect(consoleWarnSpy).not.toHaveBeenCalled();
        expect(consoleErrorSpy).not.toHaveBeenCalled();
    });

    test('should handle unknown step types with a warning', () => {
        const steps = [
            { type: "send_email", params: {} },
            { type: "fly_to_moon", params: { speed: "warp" } }, // Unknown step
            { type: "run_script", params: {} }
        ];
        const workflow = new DynamicWorkflow(steps);
        workflow.execute();

        // Check if the warning for the unknown step was logged
        expect(consoleWarnSpy).toHaveBeenCalledWith("⚠️ Unknown step type: 'fly_to_moon'. Skipping.");

        // Check that known steps still executed (based on log messages)
        const logOutput = consoleLogSpy.mock.calls.flat().join('\n');
        expect(logOutput).toContain("Simulating sending email");
        expect(logOutput).toContain("Simulating running script");

        // Check that no errors were logged
        expect(consoleErrorSpy).not.toHaveBeenCalled();
    });

     test('should use parameters passed to handlers', () => {
        const steps = [
            { type: "notify_slack", params: { channel: "#specific-channel", message: "Specific message!" } }
        ];
        const workflow = new DynamicWorkflow(steps);
        workflow.execute();

        const logOutput = consoleLogSpy.mock.calls.flat().join('\n');
        expect(logOutput).toContain("Simulating notifying Slack channel '#specific-channel': 'Specific message!...'" );
    });
});

File: creational/builder/dynamic_workflow/go/dynamic_workflow/dynamic_workflow_test.go
package dynamic_workflow

import (
	"bytes" // To capture stdout
	"io"
	"log"
	"os"
	"reflect" // For deep comparison
	"strings"
	"testing"
)

// Helper function to capture stdout
func captureOutput(f func()) string {
	var buf bytes.Buffer
	log.SetOutput(&buf)
	// Redirect standard output temporarily
	r, w, _ := os.Pipe()
	origStdout := os.Stdout
	os.Stdout = w

	f() // Execute the function whose output we want to capture

	// Restore standard output
	w.Close()
	os.Stdout = origStdout
	log.SetOutput(os.Stderr) // Restore log output

	out, _ := io.ReadAll(r)
	return string(out)
}

func TestDynamicWorkflowBuilder(t *testing.T) {
	t.Run("BuildEmptyWorkflow", func(t *testing.T) {
		builder := NewDynamicWorkflowBuilder()
		workflow := builder.Build()
		if workflow == nil {
			t.Fatal("Build() returned nil")
		}
		if len(workflow.steps) != 0 {
			t.Errorf("Expected 0 steps, got %d", len(workflow.steps))
		}
	})

	t.Run("AddSingleStep", func(t *testing.T) {
		builder := NewDynamicWorkflowBuilder()
		params := map[string]interface{}{"to": "test@example.com"}
		workflow := builder.AddStep("send_email", params).Build()
		if len(workflow.steps) != 1 {
			t.Fatalf("Expected 1 step, got %d", len(workflow.steps))
		}
		if workflow.steps[0].Type != "send_email" {
			t.Errorf("Expected step type 'send_email', got '%s'", workflow.steps[0].Type)
		}
		if !reflect.DeepEqual(workflow.steps[0].Params, params) {
			t.Errorf("Expected params %v, got %v", params, workflow.steps[0].Params)
		}
	})

	t.Run("AddMultipleStepsChaining", func(t *testing.T) {
		builder := NewDynamicWorkflowBuilder()
		params1 := map[string]interface{}{"path": "/script1"}
		params2 := map[string]interface{}{"channel": "#test"}
		workflow := builder.
			AddStep("run_script", params1).
			AddStep("notify_slack", params2).
			Build()

		if len(workflow.steps) != 2 {
			t.Fatalf("Expected 2 steps, got %d", len(workflow.steps))
		}
		if workflow.steps[0].Type != "run_script" {
			t.Errorf("Expected step 1 type 'run_script', got '%s'", workflow.steps[0].Type)
		}
		if !reflect.DeepEqual(workflow.steps[0].Params, params1) {
			t.Errorf("Expected step 1 params %v, got %v", params1, workflow.steps[0].Params)
		}
		if workflow.steps[1].Type != "notify_slack" {
			t.Errorf("Expected step 2 type 'notify_slack', got '%s'", workflow.steps[1].Type)
		}
		if !reflect.DeepEqual(workflow.steps[1].Params, params2) {
			t.Errorf("Expected step 2 params %v, got %v", params2, workflow.steps[1].Params)
		}
	})

	t.Run("AddStepReturnsSelf", func(t *testing.T) {
		builder := NewDynamicWorkflowBuilder()
		ret := builder.AddStep("test", nil)
		if ret != builder {
			t.Error("AddStep did not return the builder instance")
		}
	})
}

func TestDynamicWorkflowExecution(t *testing.T) {

	t.Run("ExecuteEmptyWorkflow", func(t *testing.T) {
		workflow := NewDynamicWorkflow([]WorkflowStep{}) // Empty steps
		output := captureOutput(workflow.Execute)

		if !strings.Contains(output, "Workflow has no steps") {
			t.Errorf("Expected output to contain 'Workflow has no steps', got '%s'", output)
		}
	})

	t.Run("ExecuteKnownSteps", func(t *testing.T) {
		steps := []WorkflowStep{
			{Type: "send_email", Params: map[string]interface{}{"to": "exec@test.com", "subject": "Exec Test"}},
			{Type: "run_script", Params: map[string]interface{}{"path": "/exec/test.py"}},
			{Type: "notify_slack", Params: map[string]interface{}{"channel": "#exec", "message": "Done"}},
		}
		workflow := NewDynamicWorkflow(steps)
		output := captureOutput(workflow.Execute)

		if !strings.Contains(output, "Simulating sending email to 'exec@test.com'") {
			t.Error("Output missing expected email simulation")
		}
		if !strings.Contains(output, "Simulating running script: '/exec/test.py'") {
			t.Error("Output missing expected script simulation")
		}
		if !strings.Contains(output, "Simulating notifying Slack channel '#exec'") {
			t.Error("Output missing expected Slack simulation")
		}
		if strings.Contains(output, "Unknown step type") {
			t.Error("Output unexpectedly contains 'Unknown step type'")
		}
		if strings.Contains(output, "Error executing step") {
			t.Error("Output unexpectedly contains 'Error executing step'")
		}
	})

	t.Run("ExecuteUnknownStep", func(t *testing.T) {
		steps := []WorkflowStep{
			{Type: "send_email", Params: map[string]interface{}{}},
			{Type: "warp_drive_engage", Params: map[string]interface{}{"speed": 9}},
			{Type: "run_script", Params: map[string]interface{}{}},
		}
		workflow := NewDynamicWorkflow(steps)
		output := captureOutput(workflow.Execute)

		if !strings.Contains(output, "Simulating sending email") {
			t.Error("Output missing expected email simulation")
		}
		if !strings.Contains(output, "⚠️ Unknown step type: 'warp_drive_engage'. Skipping.") {
			t.Error("Output missing expected unknown step warning")
		}
		if !strings.Contains(output, "Simulating running script") {
			t.Error("Output missing expected script simulation")
		}
		if strings.Contains(output, "Error executing step") {
			t.Error("Output unexpectedly contains 'Error executing step'")
		}
	})

	t.Run("HandlerParametersUsed", func(t *testing.T) {
		steps := []WorkflowStep{
			{Type: "notify_slack", Params: map[string]interface{}{"channel": "#specific-go", "message": "Go Specific Message!"}},
		}
		workflow := NewDynamicWorkflow(steps)
		output := captureOutput(workflow.Execute)

		if !strings.Contains(output, "Simulating notifying Slack channel '#specific-go': 'Go Specific Message!...'" ) {
			t.Errorf("Output missing expected Slack simulation with specific params. Got: %s", output)
		}
	})
}
